presto-main/src/main/java/com/facebook/presto/sql/planner/SystemPartitioningHandle.java

```diff
@@ -145,7 +147,7 @@ else if (partitioning == SystemPartitioning.SINGLE) {
             nodes = nodeSelector.selectRandomNodes(1);
         }
         else if (partitioning == SystemPartitioning.FIXED) {
-            nodes = nodeSelector.selectRandomNodes(getHashPartitionCount(session));
+            nodes = nodeSelector.selectRandomNodes(min(getHashPartitionCount(session), getMaxTasksPerStage(session)));
```

Comment:
You might need to change other places used `getHashPartitionCount()`. Especially  the usage in `AddExchange` and `RewriteSpatialPartitioningAggregation` .

Maybe refactor this `min` operation as a static helper method in `SystemPartitioningHandle` ?

Reply:
In fact, do we want to apply `min` here ? Today, we cannot execute when we have more system partitions than number of tasks. 

However, in the future, we might want to materialize system partitioning result (as in contrast to Hive partitioning). And at that time it would be legit to have 1000 partitions, while the max tasks per stages is only 300?

Reply:
> Especially the usage in `AddExchange`

This is used when creating connector partitioning. The number of hash partitions for connector partitioning shouldn't depend on the number of nodes. (we want to have more buckets than we have nodes)

> `RewriteSpatialPartitioningAggregation`

This is needed for partitioned spatial join. We can work on that later.

> Maybe refactor this `min` operation as a static helper method in `SystemPartitioningHandle`

It is used only in one place. Once there are more places - we can refactor it.

Reply:
You are right, this is about getting the `NodePartitioningMap`, so we should enforce `max-stages-per-task`.  While in `AddExchange` and other plan optimizer, we should only use `hash_partition_count` since it's a logical partitioning. I am assuming today we will fail when `hash_partition_count > max-stages-per-task`

Reply:
It won't fail, as we don't have to specify the number of partitions for the system partitioning upfront. System partitioning is still a special case for now.